controllers/apps/components/internal/component_base_stateful_hscale_backup.go

```diff
@@ -0,0 +1,676 @@
+package internal
+
+import (
+	"fmt"
+	"strings"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/apecloud/kubeblocks/apis/apps/v1alpha1"
+	dataprotectionv1alpha1 "github.com/apecloud/kubeblocks/apis/dataprotection/v1alpha1"
+	"github.com/apecloud/kubeblocks/internal/constant"
+	"github.com/apecloud/kubeblocks/internal/controller/builder"
+	"github.com/apecloud/kubeblocks/internal/controller/component"
+	intctrlutil "github.com/apecloud/kubeblocks/internal/controllerutil"
+)
+
+type DataClone interface {
+	Succeed() (bool, error)
+	CloneData(DataClone) ([]client.Object, error)
+	ClearTmpResources() ([]client.Object, error)
+	checkBackupStatus() (backupStatus, error)
+	backup() ([]client.Object, error)
+	checkRestoreStatus(types.NamespacedName) (backupStatus, error)
+	restore(name types.NamespacedName) ([]client.Object, error)
+}
+
+type backupStatus string
+
+const (
+	backupStatusNotCreated backupStatus = "NotCreated"
+	backupStatusProcessing backupStatus = "Processing"
+	backupStatusReadyToUse backupStatus = "ReadyToUse"
+	backupStatusFailed     backupStatus = "Failed"
+)
+
+func NewDataClone(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	cluster *appsv1alpha1.Cluster,
+	component *component.SynthesizedComponent,
+	stsObj *appsv1.StatefulSet,
+	stsProto *appsv1.StatefulSet,
+	key types.NamespacedName) DataClone {
+	if component == nil || component.HorizontalScalePolicy == nil {
+		return &baseDataClone{}
+	}
+	switch component.HorizontalScalePolicy.Type {
+	case appsv1alpha1.HScaleDataClonePolicyFromSnapshot:
+		return &snapshotDataClone{
+			baseDataClone{
+				reqCtx:    reqCtx,
+				cli:       cli,
+				cluster:   cluster,
+				component: component,
+				stsObj:    stsObj,
+				stsProto:  stsProto,
+				key:       key,
+			},
+		}
+	case appsv1alpha1.HScaleDataClonePolicyFromBackup:
+		return &backupDataClone{
+			baseDataClone{
+				reqCtx:    reqCtx,
+				cli:       cli,
+				cluster:   cluster,
+				component: component,
+				stsObj:    stsObj,
+				stsProto:  stsProto,
+				key:       key,
+			},
+		}
+	}
+	return &baseDataClone{}
+}
+
+type baseDataClone struct {
+	reqCtx    intctrlutil.RequestCtx
+	cli       client.Client
+	cluster   *appsv1alpha1.Cluster
+	component *component.SynthesizedComponent
+	stsObj    *appsv1.StatefulSet
+	stsProto  *appsv1.StatefulSet
+	key       types.NamespacedName
+}
+
+var _ DataClone = &baseDataClone{}
```

Comment:
the `baseDataClone` struct provide base struct, there is no need to comply with DataClone interface as most of the interface function is not implemented.